File: backend/app/schemas/voice_recommendation.py
"""Pydantic schemas for Phase 5 - Voice → Personalized Scheme Recommendation.

These models define the request/response contract for ``POST /voice/recommend``.

The voice layer is a thin adapter. It reuses the existing Module 4
``RecommendationMatchResponse`` schema for the returned schemes so the Flutter
client can keep parsing recommendations exactly as it does for the text-based
flow.
"""
from __future__ import annotations

from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field

from app.schemas.normalization import NormalizationResponse
from app.schemas.recommendation import RecommendationMatchResponse


class VoiceRecommendationRequest(BaseModel):
    """Request body for ``POST /voice/recommend``.

    Two mutually-supported input styles are allowed:

    1. ``text`` — a raw (or normalized) transcript. The backend will run it
       through the existing normalization service to obtain a structured query.
    2. ``normalization`` — a fully structured ``NormalizationResponse``
       (the output of ``POST /voice/normalize``). When provided, no additional
       normalization is performed.
    """

    text: Optional[str] = Field(
        default=None,
        min_length=1,
        max_length=2000,
        description="Raw or normalized voice text to interpret.",
    )
    normalization: Optional[NormalizationResponse] = Field(
        default=None,
        description="Pre-normalized structured query (output of POST /voice/normalize).",
    )
    limit: int = Field(
        default=5,
        ge=1,
        le=20,
        description="Maximum number of schemes to return.",
    )

    def model_post_init(self, __context: Any) -> None:
        """Validate that at least one input style is provided."""
        if not self.text and self.normalization is None:
            raise ValueError("Either 'text' or 'normalization' must be provided")


class VoiceRecommendationProfileView(BaseModel):
    """Minimal, verified profile view returned for a ``profile_query`` intent."""

    citizen_id: str
    full_name: Optional[str] = None
    occupation: Optional[str] = None
    annual_income: Optional[float] = None
    income_category: Optional[str] = None
    is_farmer: Optional[bool] = None
    is_disabled: Optional[bool] = None
    caste: Optional[str] = None
    community: Optional[str] = None
    education_level: Optional[str] = None
    family_member_count: Optional[int] = None
    state: Optional[str] = None
    district: Optional[str] = None
    profile_completion_percentage: Optional[int] = None


class VoiceRecommendationResponse(BaseModel):
    """Response body for ``POST /voice/recommend``.

    Reuses the existing ``RecommendationMatchResponse`` for each scheme, so the
    client's existing recommendation rendering continues to work unchanged.
    Minimal voice metadata is appended for the UI.
    """

    schemes: List[RecommendationMatchResponse] = Field(
        default_factory=list,
        description="Personalized scheme recommendations (existing schema).",
    )
    intent: str = Field(
        default="unknown",
        description="The intent that was handled.",
    )
    language: Optional[str] = Field(
        default=None,
        description="Approximate language tag of the query.",
    )
    normalized_text: Optional[str] = Field(
        default=None,
        description="Normalized representation of the citizen's query.",
    )
    confidence: float = Field(
        default=0.0,
        description="Confidence in the structured interpretation (0.0 to 1.0).",
    )
    source: Optional[str] = Field(
        default=None,
        description="Whether normalization came from llm or heuristic.",
    )
    message: Optional[str] = Field(
        default=None,
        description="Human-friendly message for unsupported intents / no results.",
    )
    profile: Optional[VoiceRecommendationProfileView] = Field(
        default=None,
        description="Verified profile returned for profile_query intents.",
    )

File: frontend/govt_scheme_app/lib/models/voice_recommendation.dart
import 'package:flutter/foundation.dart';

import 'recommendation.dart';

/// Represents the response from the backend `POST /voice/recommend` endpoint.
///
/// The backend reuses the existing `RecommendationMatchResponse` schema for
/// each scheme in `schemes`, so the client parses them with the same
/// `RecommendationMatch.fromJson` used by the text-based recommendation flow.
/// Minimal voice metadata is exposed for the UI.
@immutable
class VoiceRecommendationResult {
  const VoiceRecommendationResult({
    required this.schemes,
    required this.intent,
    this.language,
    this.normalizedText,
    this.confidence = 0.0,
    this.source,
    this.message,
    this.profile,
  });

  /// Personalized scheme recommendations (existing schema).
  final List<RecommendationMatch> schemes;

  /// The voice intent that was handled.
  final String intent;

  /// Approximate language tag of the query (e.g. "ta-en").
  final String? language;

  /// Normalized representation of the citizen's query.
  final String? normalizedText;

  /// Confidence in the structured interpretation (0.0 to 1.0).
  final double confidence;

  /// Whether normalization came from "llm" or "heuristic".
  final String? source;

  /// Human-friendly message for unsupported intents / no results.
  final String? message;

  /// Verified profile returned for `profile_query` intents.
  final VoiceProfileView? profile;

  bool get hasSchemes => schemes.isNotEmpty;

  factory VoiceRecommendationResult.fromJson(Map<String, dynamic> json) {
    return VoiceRecommendationResult(
      schemes: (json['schemes'] as List? ?? const [])
          .whereType<Map>()
          .map(
            (item) =>
                RecommendationMatch.fromJson(Map<String, dynamic>.from(item)),
          )
          .toList(),
      intent: (json['intent'] ?? 'unknown').toString(),
      language: json['language']?.toString(),
      normalizedText: json['normalized_text']?.toString(),
      confidence: double.tryParse(json['confidence']?.toString() ?? '') ?? 0.0,
      source: json['source']?.toString(),
      message: json['message']?.toString(),
      profile: json['profile'] is Map<String, dynamic>
          ? VoiceProfileView.fromJson(
              Map<String, dynamic>.from(json['profile'] as Map),
            )
          : null,
    );
  }
}

/// Minimal, verified profile view returned for a `profile_query` intent.
@immutable
class VoiceProfileView {
  const VoiceProfileView({
    required this.citizenId,
    this.fullName,
    this.occupation,
    this.annualIncome,
    this.incomeCategory,
    this.isFarmer,
    this.isDisabled,
    this.caste,
    this.community,
    this.educationLevel,
    this.familyMemberCount,
    this.state,
    this.district,
    this.profileCompletionPercentage,
  });

  final String citizenId;
  final String? fullName;
  final String? occupation;
  final double? annualIncome;
  final String? incomeCategory;
  final bool? isFarmer;
  final bool? isDisabled;
  final String? caste;
  final String? community;
  final String? educationLevel;
  final int? familyMemberCount;
  final String? state;
  final String? district;
  final int? profileCompletionPercentage;

  factory VoiceProfileView.fromJson(Map<String, dynamic> json) {
    return VoiceProfileView(
      citizenId: (json['citizen_id'] ?? '').toString(),
      fullName: json['full_name']?.toString(),
      occupation: json['occupation']?.toString(),
      annualIncome: double.tryParse(json['annual_income']?.toString() ?? ''),
      incomeCategory: json['income_category']?.toString(),
      isFarmer: json['is_farmer'] == true,
      isDisabled: json['is_disabled'] == true,
      caste: json['caste']?.toString(),
      community: json['community']?.toString(),
      educationLevel: json['education_level']?.toString(),
      familyMemberCount: int.tryParse(
        json['family_member_count']?.toString() ?? '',
      ),
      state: json['state']?.toString(),
      district: json['district']?.toString(),
      profileCompletionPercentage: int.tryParse(
        json['profile_completion_percentage']?.toString() ?? '',
      ),
    );
  }
}
